zrml/prediction-markets/src/lib.rs

```diff
@@ -139,27 +151,47 @@ mod pallet {
             let market_status = market.status;
             let outcome_assets = Self::outcome_assets(market_id, &market);
             let outcome_assets_amount = outcome_assets.len();
-            Self::clear_auto_resolve(&market_id)?;
-            T::MarketCommons::remove_market(&market_id)?;
-            Self::deposit_event(Event::MarketDestroyed(market_id));
+            let market_account = Self::market_account(market_id);
 
-            let mut outcome_assets_iter = outcome_assets.into_iter();
+            // Slash outstanding bonds; see
+            // https://github.com/zeitgeistpm/runtime-audit-1/issues/34#issuecomment-1120187097 for
+            // details.
+            let slash_market_creator = |amount| {
+                CurrencyOf::<T>::slash_reserved_named(&RESERVE_ID, &market.creator, amount);
+            };
+            if market_status == MarketStatus::Proposed {
+                slash_market_creator(T::AdvisoryBond::get());
+            }
+            if market_status != MarketStatus::Resolved
+                && market_status != MarketStatus::InsufficientSubsidy
+            {
+                if market.creation == MarketCreation::Permissionless {
+                    slash_market_creator(T::ValidityBond::get());
+                }
+                slash_market_creator(T::OracleBond::get());
+            }
 
-            // Delete of this market's outcome assets.
-            let mut manage_outcome_asset = |asset: Asset<_>| -> usize {
+            // Delete market's outcome assets, clear market and delete pool if necessary.
+            let mut destroy_asset = |asset: Asset<_>| -> usize {
                 let (total_accounts, accounts) = T::Shares::accounts_by_currency_id(asset);
                 share_accounts = share_accounts.saturating_add(accounts.len());
                 T::Shares::destroy_all(asset, accounts.iter().cloned());
                 total_accounts
             };
-
-            if let Some(first_asset) = outcome_assets_iter.next() {
-                total_accounts = manage_outcome_asset(first_asset);
+            for asset in outcome_assets.into_iter() {
+                total_accounts = destroy_asset(asset);
             }
-            for asset in outcome_assets_iter {
-                let _ = manage_outcome_asset(asset);
+            CurrencyOf::<T>::slash(&market_account, CurrencyOf::<T>::free_balance(&market_account));
+            if let Ok(pool_id) = T::MarketCommons::market_pool(&market_id) {
+                T::Swaps::destroy_pool(pool_id)?;
```

Comment:
The `#[transactional]` is added later.

Other than that, the reasoning for checking for errors is the same a the reasoning for not allowing anything that panicks in our code: Maybe someone changes `destroy_pool` in the future so that it can return an error that must be handled by `admin_destroy_market`, but overlooks that `admin_destroy_market` must be changed, as well.